source/s3_client.c

```diff
@@ -1130,6 +1230,54 @@ static void s_s3_client_schedule_process_work_synced_default(struct aws_s3_clien
     client->synced_data.process_work_task_scheduled = true;
 }
 
+/* Task function for trying to find a request that can be processed. */
+static void s_s3_client_trim_buffer_pool_task(struct aws_task *task, void *arg, enum aws_task_status task_status) {
+    AWS_PRECONDITION(task);
+    (void)task;
+    (void)task_status;
+
+    /* Client keeps a reference to the event loop group; a 'canceled' status should not happen.*/
+    AWS_ASSERT(task_status == AWS_TASK_STATUS_RUN_READY);
+
+    struct aws_s3_client *client = arg;
+    AWS_PRECONDITION(client);
+
+    uint32_t num_reqs_in_flight = (uint32_t)aws_atomic_load_int(&client->stats.num_requests_in_flight);
```

Comment:
still iffy about using `stats`. why not check `aws_linked_list_empty(&client->threaded_data.meta_requests)`?

Reply:
threaded.meta_requests can have a subset of requests, since there could be some in the synced.pending queue. stats one seemed easier to use and we already rely on it in scheduling and logic for it is fairly simple.